### PR TITLE
Improve geometry nan support / default ZM support

### PR DESCRIPTION
**Pull request by m-kuhn**
## Description

- We export `nan` into wkt but did not support parsing the same -> we now support nan for Z and M in wkt parsing
- When upcasting geometries to Z/M counterparts, 0.0 was hardcoded value. This extends the API to provide default values for these dimensions

### CODE REVIEW

**Comment by rouault** [comment]
> ##
> 
>     * We export `nan` into wkt

just to note that this is an extension to the WKT spec and that neither OGR or PostGIS supports reading that.

**Comment by m-kuhn** [comment]
I am aware of that.
Is there a more correct way to store NULL values for Z into a 3D table?

**Comment by 3nids** [comment]
This is conflicting with the approach used in another PR https://github.com/qgis/QGIS/pull/46687
see https://github.com/qgis/QGIS/pull/46687/commits/d5bd76e6222fedfe8afdeac7e658cfbcaf5bcaf4
image

**Comment by 3nids** [comment]
shall we use `std::numeric_limits<float>::min()` for the default args so we could have both approaches?

**Comment by m-kuhn** [comment]
I don't think it's conflicting. This is offering API that can be connected to these settings values but ultimately leaves control to the consumer of the API (e.g. a python plugin can provide something else than those user/profile defined settings).

**Comment by 3nids** [comment]
OK, I'll change the default values from the call to the method whenever this is merged.

**Comment by 3nids** [comment]
But, wouldn't it make more sense that the defaults are NaN rather than 0? 
ping @lbartoletti

**Comment by nyalldawson** [comment]
> just to note that this is an extension to the WKT spec and that neither OGR or PostGIS supports reading that.

And it causes a lot of annoying bugs as a result :fearful:

**Comment by m-kuhn** [comment]
This PR does not export more nan. It just brings our parsing capabilities on par with what we export already. So it effectively reduces those bugs. Worst case this is dead code at some point because no one exports nan any more 😅 

I'd still be interested in hearing proposals for a more correct z null handling.

**Comment by m-kuhn** [comment]
@nyalldawson are you ok with this fix or would you prefer a different approach?

File: tests/src/python/test_qgsgeometry.py [hunk 1]
```diff
@@ -6292,6 +6292,13 @@ def coerce_to_wkt(wkt, type, defaultZ=None, defaultM=None):
                                        QgsWkbTypes.MultiLineString),
                          ['MultiLineString ((1 1, 1 2, 2 2, 1 1),(3 3, 4 3, 4 4, 3 3))'])
 
+        # Missing z values
+        self.assertEqual(coerce_to_wkt('Point (1 1)', QgsWkbTypes.PointZ), ['PointZ (1 1 nan)'])
```
Comment (strk) [inline tests/src/python/test_qgsgeometry.py:6296]: For the record, PostGIS uses 0 as the missing value when forcing dimensions:
```
strk=# select ST_AsEWKT(ST_Force3DZ('POINT(1 1)'::geometry)) force_z, ST_AsEWKT(ST_Force3DM('POINT(1 1)'::geometry)) force_m;
   force_z    |    force_m    
--------------+---------------
 POINT(1 1 0) | POINTM(1 1 0)
(1 row)
```
PostGIS will consider geometries with NaN values as invalid:
```
strk=# select valid, reason, ST_AsEWKT(location)  FROM ST_IsValidDetail('POINT(0 NaN 0)'::geometry);
 valid |       reason       |   st_asewkt    
-------+--------------------+----------------
 f     | Invalid Coordinate | POINT(0 NaN 0)
(1 row)
```

PostGIS will NOT catch NaN in Z as being invalid (but I've no idea how many warms will be out their cans):
```
strk=# select valid, reason, ST_AsEWKT(location)  FROM ST_IsValidDetail('POINT(0 0 NaN)'::geometry);
 valid | reason | st_asewkt 
-------+--------+-----------
 t     |        | 
(1 row)
```

All of the above happens with POSTGIS="3.3.0dev 3.2.0-347-gf26b2ce71"

**Comment by nyalldawson** [comment]
@m-kuhn 
> are you ok with this fix or would you prefer a different approach?

I've nothing against this fix (since it makes **reading** more robust), but I do think it was a mistake to introduce nan support in z/m values in the first place and I believe that we should remove support for writing these (at least for the OGR and Postgres providers, possibly other providers too). I've run into many customer issues with datasets/workflows which break where the fault has eventually been tracked down to nan z values in a feature. (On both OGR and postgres datasets, and on custom workflows where geometries are exported as WKT and then sent to some other API).

**Comment by nyalldawson** [comment]
@rouault 
> I believe that we should remove support for writing these (at least for the OGR ... provider)

What's your thoughts here? Is writing nan values to OGR feature coordinates something that is ever safe to do?

**Comment by rouault** [comment]
> What's your thoughts here? Is writing nan values to OGR feature coordinates something that is ever safe to do?

It is difficult to assess globally how all parts of OGR would behave when feed with NaN coordinates... OGRPoint is ready for that for the POINT EMPTY representation from WKB. For other cases, this must be mostly untested.

Perhaps instead of NaN,  -DOUBLE_MAX could be used ? The good old shapefile spec says "Floating point numbers must be numeric values. Positive infinity, negative infinity, and
Not-a-Number (NaN) values are not allowed in shapefiles. Nevertheless, shapefiles
support the concept of "no data" values, but they are currently used only for measures.
Any floating point number smaller than –1e38 is considered by a shapefile reader to
represent a "no data" value."

**Comment by m-kuhn** [comment]
I have also experienced such issues and this is an attempt to improve things.

Keeping nan as some magic value (-DOUBLE_MAX) inside QGIS (and cosmetically improving it e.g. in the node tool) is a good option. It will still have issues with things like calculating the avg height of features and the readability of wkt.

If external interfaces are the main concern here, we could adjust things there
- In providers: map nan to -DOUBLE_MAX
- In wkt export: cut dimensions with nan (`PointZ(1 2 nan)` -> `Point(1 2)`)

This would still allow us to be "correct" inside QGIS but vastly improve the compatibility with the outside world. And if a provider supports a different concept for null z we can adjust the compatibility interface specifically.

File: tests/src/python/test_qgsgeometry.py (hunk 1 above)
Comment (lbartoletti) [reply tests/src/python/test_qgsgeometry.py:6296]: And Postgis doesn't allow NaN in WKT (at reading):

```sql
# SELECT St_AsText(ST_MakePoint(1, 2, 'NaN'));
     st_astext
-------------------
 POINT Z (1 2 NaN)
```

But:

```sql
# SELECT ST_GeomFromText('POINTZ(1 2 NaN)');
ERREUR:  parse error - invalid geometry
ASTUCE : "POINTZ(1 2 NaN)" <-- parse error at position 15 within geometry

# SELECT St_AsText(ST_GeomFromText(ST_AsText(ST_MakePoint(1, 2, 'NaN'))));
ERREUR:  parse error - invalid geometry
ASTUCE : "POINT Z (1 2 NaN)" <-- parse error at position 17 within geometry
```

**Comment by lbartoletti** [comment]
I think there is a lack of information into the OGC spec. But I agree that NaN should only be available for `POINT EMPTY` and I like the Shapefile spec: no NaN or +/-Inf.

BUT, QGIS doesn't constraints the input of the coordinates. We have started to add constraints, but there is still work to do (and it could break things IIRC).

Today, a QgsPoint will return "EMPTY" if X and Y are NaN:

```python
p = QgsPoint(math.nan, math.nan, math.nan, math.nan, QgsWkbTypes.PointZM)
p.asWkt()
#'PointZM EMPTY'
p.setX(1)
p.asWkt()
#'PointZM EMPTY'
p.setY(2)
p.asWkt()
#'PointZM (1 2 nan nan)'
```
So, Is the last line should be `Point(1 2)`? And after:

```python
p.setM(4)
p.asWkt()
#'PointZM (1 2 nan 4)'
```
Is p should be `PointM (1 2 4)`?


And for the other geometries?

```python
l = QgsLineString([(1, math.nan), (2, 3)])
l.asWkt()
# 'LineString (1 nan, 2 3)'
l = QgsLineString([(1, 3, math.nan), (2, 3, 5)])
l.asWkt()
# 'LineStringZ (1 3 nan, 2 3 5)'
l = QgsLineString([(math.nan, math.nan), (math.nan, math.nan)])
l.asWkt()
# 'LineString (nan nan, nan nan)'
```

Is it acceptable? I guess not.

**Comment by nyalldawson** [comment]
@m-kuhn 
> If external interfaces are the main concern here, we could adjust things there
> In providers: map nan to -DOUBLE_MAX

I agree that this is a good approach -- then we can adjust individual providers depending on what value makes most sense for that provider (e.g. if a provider DOES support nan z/m we can leave things unchanged).

We could potentially also do the opposite when reading features in for that provider, and eg for OGR provider automatically flip -DOUBLE_MAX z/m values to nan.

> In wkt export: cut dimensions with nan (PointZ(1 2 nan) -> Point(1 2))

This one concerns me a little, in that the round trip geom->wkt->geom we result in a different dimensioned geometry. I think a more flexible approach might be to add a flags argument to QgsAbstractGeometry/QgsGeometry::asWkt to allow the caller to control the behavior, with flags for dropping dimensions, using nan, or using -doublemax (and we could default to -doublemax)

**Comment by m-kuhn** [comment]
That sounds good to me.

Dropping the dimensions might be a bad idea, as @lbartoletti pointed out correctly with `LineStringZ(1 2 3, 3 2 nan, 4 5 6)` will potentially drop valuable information, so I'd say we just stick to either nan or -double::max.

Thanks everyone for all the good feedback in here.

**Comment by m-kuhn** [comment]
I do want to continue this, but will get to it June earliest

**Comment by m-kuhn** [comment]
I think this is worth merging as it improves reading nan from wkt support.

**Comment by nirvn** [comment]
I would very much like to see this merged for 3.30 (and backported to 3.28). Getting null geometry from a `PointMZ( 1 2 nan nan )` is quite bad and this PR fixes that.

File: src/core/geometry/qgspoint.cpp [hunk 2]
```diff
@@ -154,6 +154,14 @@ bool QgsPoint::fromWkb( QgsConstWkbPtr &wkbPtr )
   return true;
 }
 
+inline double stringToDouble( const QString &string )
+{
+  if ( 0 == string.compare( QLatin1String( "nan" ) ) )
```
Comment (3nids) [inline src/core/geometry/qgspoint.cpp:159]:
```suggestion
  if ( 0 == string.compare( QLatin1String( "nan" ), Qt::CaseInsensitive ) )
```

File: tests/src/python/test_qgsgeometry.py [hunk 3]
```diff
@@ -6420,6 +6420,12 @@ def coerce_to_wkt(wkt, type, defaultZ=None, defaultM=None):
...
```
Comment (m-kuhn) [inline tests/src/python/test_qgsgeometry.py:6424]:
```suggestion
#        self.assertEqual(coerce_to_wkt('Point (1 1)', QgsWkbTypes.PointZ), ['PointZ (1 1 nan)'])
```

Let's comment that for noe

**Review by strk — APPROVED** [review]
Sounds good to me, thanks !

**Comment by strk** [comment]
This is the CI issue:
```
======================================================================
FAIL: testCoerce (__main__.TestQgsGeometry)
Test coerce function
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/QGIS/tests/src/python/test_qgsgeometry.py", line 6426, in testCoerce
    self.assertEqual(coerce_to_wkt('PointZ (1 1 nan)', QgsWkbTypes.PointZM), ['PointZM (1 1 nan nan)'])
AssertionError: Lists differ: [] != ['PointZM (1 1 nan nan)']

Second list contains 1 additional elements.
First extra element 0:
'PointZM (1 1 nan nan)'

- []
+ ['PointZM (1 1 nan nan)']

----------------------------------------------------------------------
```

Should be fixed before merge.
Please also consider adding more tests with ZM inputs and Z-only and M-only outputs

**Pull request closed without merging**